resolver/master_changer.go

```diff
@@ -0,0 +1,133 @@
+package resolver
+
+import (
+	"encoding/binary"
+	"net"
+	"sort"
+	"strconv"
+
+	mesos "github.com/mesos/mesos-go/mesosproto"
+
+	"github.com/mesosphere/mesos-dns/logging"
+)
+
+// MasterChanger stores the current state
+// of the current leader/masters
+type MasterChanger struct {
+	// current masters list,
+	// 1st item represents the leader,
+	// the rest remaining masters
+	masters []string
+
+	changed chan []string
+}
+
+// NewMasterChanger returns a new MasterChanger
+// with an unknown (empty) leader, the given masters
+// and the initial channel capacity for emitting changes
+func NewMasterChanger(masters []string, cap int) *MasterChanger {
+	m := make([]string, 1, len(masters)+1)
+	m[0] = ""
+	m = append(m, masters...)
+
+	return &MasterChanger{
+		masters: m,
+		changed: make(chan []string, cap),
+	}
+}
+
+// ChangedChan returns a channel of masters.
+//
+// The emitted string slice follows the following convention:
+// - The first item of the slice represents the leader
+// - The remaining items represent additional masters
+//
+// Note: subsequent calls to ChangedChan return the same channel
+func (mc *MasterChanger) ChangedChan() <-chan []string {
+	return mc.changed
+}
+
+// OnMasterChanged sets the given MasterInfo as the current leader
+// leaving the remaining masters unchanged and emits the current masters state
+func (mc *MasterChanger) OnMasterChanged(info *mesos.MasterInfo) {
+	logging.VeryVerbose.Println("Updated leader: ", info)
+
+	if info == nil {
+		logging.Error.Println("No master available in Zookeeper.")
+		return
+	}
+
+	master := masterHostPort(info)
+	mc.set(master, mc.masters[1:])
+	mc.Emit()
+}
+
+// UpdatedMasters sets the given slice of MasterInfo as the current remaining masters
+// leaving the current leader unchanged and emits the current masters state
+func (mc *MasterChanger) UpdatedMasters(infos []*mesos.MasterInfo) {
+	logging.VeryVerbose.Println("Updated masters: ", infos)
+
+	if infos == nil {
+		logging.Error.Println("No masters available in Zookeeper.")
+		return
+	}
+
+	masters := make([]string, 0, len(infos))
+	for _, info := range infos {
+		if validMasterInfo(info) {
+			masters = append(masters, masterHostPort(info))
+		}
+	}
+
+	if len(masters) == 0 {
+		logging.Error.Println("No valid masters available in Zookeeper.")
```

Comment:
In a distant future, far far away, logging won't be mingled with business logic. :D